2022-09-28-introduce-q-d.md

```diff
@@ -0,0 +1,241 @@
+
+Dynamic Logic is a form of _modal logic_, similar to [temporal logic](https://en.wikipedia.org/wiki/Temporal_logic), but whose modalities are the _actions_ (or programs) themselves. One can intertwine _programs_ and logic predicates to specify the behaviour of the former when executing some statements and actually [Dynamic Logic](https://en.wikipedia.org/wiki/Dynamic_logic_(modal_logic)) evolved from _Hoare's Triples_.
+
+Here is the dynamic logic reformulation of the previously stated Hydra property which has been kept as close as possible to the original English statement:
+
+```haskell
+conflictFreeLiveness :: DL WorldState ()
+conflictFreeLiveness = do
+  anyActions_
+  getModelStateDL >>= \case
+    st@WorldState{hydraState = Open{}} -> do
+      (party, payment) <- forAllQ (nonConflictingTx st)
+      action $ Model.NewTx party payment
+      eventually (ObserveConfirmedTx payment)
+    _ -> pass
+ where
+  nonConflictingTx st = withGenQ (genPayment st) (const [])
+  eventually a = action (Wait 10) >> action a
+```
+
+Note that in order to define this property we have introduced two "pseudo-actions" in the _Model_, `Wait` and `ObserveConfirmedTx`: Those `Action`s have no effect on the model itself, the former being used to introduce some delay in the context of distributed and asynchronous execution, and the latter serving the purpose of _observing_ the current state of the SUT. An alternative formulation would have been to make `ObserveConfirmedTx` return the set of confirmed transactions and then express the condition as a logic predicate within the `conflictFreeLiveness` property's body.
+
+### Checking Properties
+
+The last step in putting quickcheck-dynamic at work is to be able to connect the _StateModel_, the _RunModel_, and the _DynamicLogic_ expression and turn those into a QuickCheck `Property` which can then be checked using the standard testing framework.
+
+quickcheck-dynamic provides 2 functions for that purpose. The `forAllDL_` function (actually more a family of functions) will leverage `DL` formulae to generate sequences of `Action`s:
+
+```haskell
+prop_checkConflictFreeLiveness :: Property
+prop_checkConflictFreeLiveness =
+  forAllDL_ conflictFreeLiveness prop_HydraModel
+```
+
+The `runActions` function will execute the generated trace against the `RunModel`.
+
+```haskell
+prop_HydraModel :: Actions WorldState -> Property
+prop_HydraModel actions = property $
+  runIOSimProp $ do
+    _ <- runActions runIt actions
+    assert True
+```
+
+In this particular instance from Hydra, we need some additional machinery (the `runIOSimProp` function) to handle the execution of some monadic `PropertyM` into `IOSim` monad, turning it into a `Property`.
```

Comment:
And also running the `StateT Nodes` monad transformer. But probably fine to gloss over this.

Reply:
Yeah, it's already quite complicated I guess :)